Introducing pgroll blogpost (#127)
xataio/mdx-blog: introducing-pgroll.mdx

- **Multiple steps**: Database migrations are rarely a one-and-done affair. They often involve a series of intricate steps that need to be executed meticulously. Managing these multiple steps is usually not part of the team deployment workflow, and can quickly become a logistical nightmare. ie: It can take [6 steps to rename a column](https://xata.io/blog/postgres-schema-changes-pita#application-deploys-and-the-6-stages-of-rename) without downtime!
- **Unexpected database locks**: Traditional migration methods can cause unexpected database locks, bringing services to a grinding halt and causing application downtime.
- **No easy rollbacks**: In the world of schema migrations, the safety net is often full of holes. Rolling back to a previous state in the event of a migration gone awry is rarely straightforward and is frequently a risky endeavor.

Due to these issues, many developers choose to avoid complex migrations and only make additive changes. This leads to the accumulation of **technical debt in the database schema**, such as orphaned columns or missing constraints.

## Introducing pgroll

<div style="text-align: center">

>   <img src="https://raw.githubusercontent.com/xataio/mdx-blog/introducing-pgroll/images/pgroll-banner.png" alt="pgroll banner" />

Author: ```suggestion
  <img src="https://raw.githubusercontent.com/xataio/mdx-blog/main/images/pgroll-banner.png" alt="pgroll banner" />
```

---

> title: 'Introducing pgroll, zero-downtime, undoable, schema migrations for Postgres'

Reviewer: ```suggestion
title: 'Introducing pgroll: zero-downtime, undoable, schema migrations for Postgres'
```

> title: 'Introducing pgroll, zero-downtime, undoable, schema migrations for Postgres'

Author: I'm a bit on the fence about reversible vs undoable, I checked for instance this example from flyway where they use undoable https://documentation.red-gate.com/fd/tutorial-undo-migrations-184127627.html

> title: 'Introducing pgroll, zero-downtime, undoable, schema migrations for Postgres'

Reviewer 2: I think for me, it's the word undoable for some reason -- not the act of undo-ing. The definition itself is kind of funny.

> "impossible to do : not doable. a theoretical approach that proved to be undoable in practice. 2. : able to be reversed or undone : possible to undo. Nearly everything you can do to an image is undoable.

PlanetScale also seems to have a similar "undo" feature (https://thenewstack.io/planetscale-rewind-an-undo-button-for-bad-schema-migrations/). Reversible schema migrations seems to provide similar search results, with Cassandra using this term too.

Let's leave it as undo, I'll update the newsletter. What do you think of any of these title updates to include OSS?

> pgroll: An open source tool for undoable, zero-downtime schema migrations with Postgres

> Introducing pgroll, an open source tool for undoable, zero-downtime schema migrations with Postgres

- **Multiple steps**: Database migrations are rarely a one-and-done affair. They often involve a series of intricate steps that need to be executed meticulously. Managing these multiple steps is usually not part of the team deployment workflow, and can quickly become a logistical nightmare. ie: It can take [6 steps to rename a column](https://xata.io/blog/postgres-schema-changes-pita#application-deploys-and-the-6-stages-of-rename) without downtime!
- **Unexpected database locks**: Traditional migration methods can cause unexpected database locks, bringing services to a grinding halt and causing application downtime.
- **No easy rollbacks**: In the world of schema migrations, the safety net is often full of holes. Rolling back to a previous state in the event of a migration gone awry is rarely straightforward and is frequently a risky endeavor.

Due to these issues, many developers choose to avoid complex migrations and only make additive changes. This leads to the accumulation of **technical debt in the database schema**, such as orphaned columns or missing constraints.

## Introducing pgroll

<ArticleImage

>   src="https://github.com/xataio/mdx-blog/raw/introducing-pgroll/images/pgroll-banner.png"

Author: ```suggestion
  src="https://github.com/xataio/mdx-blog/raw/main/images/pgroll-banner.png"
```

- Backfill all existing values from the old column into the new one, upgrading the data in the cases where the value was `NULL`, based on the user-defined function. Of course, backfill happens in batches, avoiding excessively large updates that could also block the database.
- Set up a trigger function so new inserts & updates get automatically copied between the new and the old columns.
- Keep the old column working at all times, ensuring the migration can be rolled back instantly if needed.
- Create a new view of the schema that hides the old column and promotes the new one in its place, exposing this way the new version of the schema while the old one keeps working.

Typically, all of these steps would require manual execution to achieve a complex update like this. However, `pgroll` streamlines the process by performing all of them with a single command ✨.

Once you are happy with the new version of the schema and all clients have been updated to use it, the old schema will no longer be accessed. It’s time to complete the migration (contract the schema) to get rid of the old column & triggers:

```json
pgroll complete
```

> For more complex examples, there is a wide range of schema migration samples in our [examples](https://github.com/xataio/pgroll/tree/main/examples) and [docs](https://github.com/xataio/pgroll/tree/doc/document-alter-column-operation/docs).

Reviewer 2: The URL seemed to break

```suggestion
For more complex examples, there is a wide range of schema migration samples in our [examples](https://github.com/xataio/pgroll/tree/main/examples) and [docs](https://github.com/xataio/pgroll/tree/main/docs).
```

## What’s next?

> Today, we are rolling out the first version of `pgroll` and we are looking forward to your feedback!  We plan to continue developing `pgroll` and exploring how it can make it easier and safer for Xata users to interact with their schema.

Reviewer 2: Small suggestion

```suggestion
Today, we are rolling out the first version of `pgroll` and we are looking forward to your feedback!  We plan to continue developing `pgroll` and exploring how it can make it easier and safer for Xata users to evolve their schema.
```

> If you have any suggestions, questions, open an issue in our [GitHub repo](https://github.com/xataio/pgroll), reach out to us on [Discord](https://xata.io/discord) or follow us on [Twitter](https://twitter.com/xata). We'd love to hear from you and keep you updated on the latest at Xata.

Reviewer 2: ```suggestion
If you have any suggestions, questions, open an issue in our [GitHub repo](https://github.com/xataio/pgroll), reach out to us on [Discord](https://xata.io/discord) or follow us on [X / Twitter](https://twitter.com/xata). We'd love to hear from you and keep you updated on the latest at Xata.
```

> If you have any suggestions, questions, open an issue in our [GitHub repo](https://github.com/xataio/pgroll), reach out to us on [Discord](https://xata.io/discord) or follow us on [Twitter](https://twitter.com/xata). We'd love to hear from you and keep you updated on the latest at Xata.

Reviewer 2: Small suggestion

```suggestion
If you have any suggestions or questions, please open an issue in our [GitHub repo](https://github.com/xataio/pgroll), reach out to us on [Discord](https://xata.io/discord) or follow us on [Twitter](https://twitter.com/xata). We'd love to hear from you and keep you up to date with the latest progress on `pgroll`.
```

> Let’s start rolling!

Reviewer 2: Still no meme!? 😆

Author: I'm open to ideas, I'm so bad at memes 😸

Reviewer 2: Honestly, @tsg is the meme master. There's lots of opportunity for this one I think. It's perfectly fine without it though haha, I'm just messing you with

![image](https://github.com/xataio/mdx-blog/assets/25181823/b50ed6c9-ca64-4bef-999f-d71974580e5b)

- **Multiple steps**: Database migrations are rarely a one-and-done affair. They often involve a series of intricate steps that need to be executed meticulously. Managing these multiple steps is usually not part of the team deployment workflow, and can quickly become a logistical nightmare. ie: It can take [6 steps to rename a column](https://xata.io/blog/postgres-schema-changes-pita#application-deploys-and-the-6-stages-of-rename) without downtime!
- **Unexpected database locks**: Traditional migration methods can cause unexpected database locks, bringing services to a grinding halt and causing application downtime.
- **No easy rollbacks**: In the world of schema migrations, the safety net is often full of holes. Rolling back to a previous state in the event of a migration gone awry is rarely straightforward and is frequently a risky endeavor.

Due to these issues, many developers choose to avoid complex migrations and only make additive changes. This leads to the accumulation of **technical debt in the database schema**, such as orphaned columns or missing constraints.

## Introducing pgroll

<ArticleImage

>   src="https://github.com/xataio/mdx-blog/raw/introducing-pgroll/images/introducing-pgroll-banner@2x.jpg"

Author: ```suggestion
  src="https://github.com/xataio/mdx-blog/raw/main/images/introducing-pgroll-banner@2x.jpg"
```
